### PR TITLE
[ISSUE #5012] support escaping transactional messages

### LINKED ISSUE #5012: Support escaping transactional messages in slave-acting-master mode
**FEATURE REQUEST**

In [RIP-32](https://github.com/apache/rocketmq/wiki/RIP-32-Slave-Acting-Master-Mode), Slave Acting Master Mode is supported and the escape of the transactional messages need to be supported. 

Below is a simple implementation of escaping transactional messages:

If a slave node is acting as master,
- EndTransaction requests (including commit and rollback) are denied.
- When checking status of half messages, escape half messages to an available master directly, without checking transaction status actually. There is a small trick: If a half message is sent to a local master, just append it to half topic; if sent to a remote master, the half message should be restored to original real message, because sending messages to a system topic is forbidden.

This implemetation keeps read-only sematics of slaves, so that no new half messages or op messages will be created in slave acting as master, all actual checking processes of half messages are delivered to master.

Previously there are several proposals, such as #4427, #4477, etc. They do provide good references and I truly appreciate the authors.

### PR DESCRIPTION
**Pull request by caigy**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

fix #5012 

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java [hunk 1]
```diff
@@ -187,6 +190,39 @@ public void check(long transactionTimeout, int transactionCheckMax,
...
+
+                            if (isSuccess) {
+                                escapeFailCnt = 0;
+                                newOffset = i + 1;
+                                i++;
+                            } else {
+                                System.out.printf("Escaping transactional message failed %s times! msgId(offsetId)=%s, UNIQ_KEY(transactionId)=%s",
+                                    escapeFailCnt + 1,
+                                    msgExt.getMsgId(),
+                                    msgExt.getUserProperty(MessageConst.PROPERTY_UNIQ_CLIENT_MESSAGE_ID_KEYIDX));
```
Comment (RongtongJin) [inline broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java:209]: We should remove the useless System.out.printf

File: test/src/test/java/org/apache/rocketmq/test/container/ContainerIntegrationTestBase.java [hunk 2]
```diff
@@ -186,6 +187,7 @@ protected static void createTopicTo(BrokerController masterBroker, String topicN
         try {
             TopicConfig topicConfig = new TopicConfig(topicName, rqn, wqn, 6, 0);
             defaultMQAdminExt.createAndUpdateTopicConfig(masterBroker.getBrokerAddr(), topicConfig);
+            System.out.println("Create topic " + topicName + " to " + masterBroker.getBrokerIdentity().getCanonicalName() + "@" + masterBroker.getBrokerAddr());
```
Comment (RongtongJin) [inline test/src/test/java/org/apache/rocketmq/test/container/ContainerIntegrationTestBase.java:190]: We should remove the useless System.out.println

File: broker/src/test/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.rocketmq.common.message.MessageConst;
+import org.apache.rocketmq.common.message.MessageExt;
+import org.apache.rocketmq.common.message.MessageExtBrokerInner;
+import org.apache.rocketmq.common.sysflag.MessageSysFlag;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class TransactionalMessageUtilTest {
```
Comment (RongtongJin) [inline broker/src/test/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageUtilTest.java:30]: How about ignoring these tests? It proves the function's correctness, but it takes too much time and is not the core function, which affects the running time of the CI. IMO, just run these tests manually before the release.

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java [hunk 4]
```diff
@@ -187,6 +190,39 @@ public void check(long transactionTimeout, int transactionCheckMax,
...
+                        ) {
+                            final MessageExtBrokerInner msgInner = this.transactionalMessageBridge.renewHalfMessageInner(msgExt);
+                            final boolean isSuccess = this.transactionalMessageBridge.escapeMessage(msgInner);
+
+                            if (isSuccess) {
+                                escapeFailCnt = 0;
+                                newOffset = i + 1;
+                                i++;
+                            } else {
+                                System.out.printf("Escaping transactional message failed %s times! msgId(offsetId)=%s, UNIQ_KEY(transactionId)=%s",
```
Comment (ShannonDing) [inline broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java:206]: better to remove stdout.

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java [hunk 5]
```diff
@@ -187,6 +190,39 @@ public void check(long transactionTimeout, int transactionCheckMax,
                             }
                         }
 
+                        if (BrokerRole.SLAVE.equals(this.transactionalMessageBridge.getBrokerController().getMessageStoreConfig().getBrokerRole())
+                            && this.transactionalMessageBridge.getBrokerController().getMinBrokerIdInGroup()
+                            == this.transactionalMessageBridge.getBrokerController().getBrokerIdentity().getBrokerId()
+                            && this.transactionalMessageBridge.getBrokerController().getBrokerConfig().isEnableSlaveActingMaster()
```
Comment (ShannonDing) [inline broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java:196]: “EnableSlaveActingMaster” is default set to false. maybe it's better to let this condition checked at first.

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java (hunk 1 above)
Comment (caigy) [reply broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java]: removed

File: test/src/test/java/org/apache/rocketmq/test/container/ContainerIntegrationTestBase.java (hunk 2 above)
Comment (caigy) [reply test/src/test/java/org/apache/rocketmq/test/container/ContainerIntegrationTestBase.java]: done

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java (hunk 4 above)
Comment (caigy) [reply broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java]: removed

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java (hunk 5 above)
Comment (caigy) [reply broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageServiceImpl.java]: new order:

1. check EnableSlaveActingMaster first for default false;
2. check brokerId for only 1 broker meeting this condition in a cluster;
3. check broker role.

File: broker/src/main/java/org/apache/rocketmq/broker/failover/EscapeBridge.java [hunk 6]
```diff
@@ -112,15 +113,21 @@ public PutMessageResult putMessage(MessageExtBrokerInner messageExt) {
     }
 
     private SendResult putMessageToRemoteBroker(MessageExtBrokerInner messageExt) {
-        final TopicPublishInfo topicPublishInfo = this.brokerController.getTopicRouteInfoManager().tryToFindTopicPublishInfo(messageExt.getTopic());
+        final boolean isTransHalfMessage = TransactionalMessageUtil.buildHalfTopic().equals(messageExt.getTopic());
+        MessageExtBrokerInner messageToPut = messageExt;
+        if (isTransHalfMessage) {
+            messageToPut = TransactionalMessageUtil.buildTransactionalMessageFromHalfMessage(messageExt);
+        }
```
Comment (RongtongJin) [inline broker/src/main/java/org/apache/rocketmq/broker/failover/EscapeBridge.java:120]: How about moving this logic to the upper layer to make putMessageToRemoteBroker function purer？

**Review by RongtongJin — DISMISSED** [review]

**Review by ShannonDing — DISMISSED** [review]

File: broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageBridge.java [hunk 7]
```diff
@@ -346,4 +346,16 @@ public MessageExt lookMessageByOffset(final long commitLogOffset) {
     public BrokerController getBrokerController() {
         return brokerController;
     }
+
+    public boolean escapeMessage(MessageExtBrokerInner messageInner) {
+        PutMessageResult putMessageResult = this.brokerController.getEscapeBridge().putMessage(messageInner);
+        if (putMessageResult != null
```
Comment (Oliverwqcwrw) [inline broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageBridge.java:352]: how about use `org.apache.rocketmq.store.PutMessageResult#isOk` to verify whether success,
It's a little bit cleaner and it's consistent with everything else

Comment (caigy) [reply broker/src/main/java/org/apache/rocketmq/broker/transaction/queue/TransactionalMessageBridge.java]: Good idea~ Changed as you advised.

**Review by Oliverwqcwrw — APPROVED** [review]

**Review by RongtongJin — APPROVED** [review]

**Pull request merged**